metrics-core/src/main/java/com/yammer/metrics/stats/ExponentiallyDecayingSample.java

```diff
@@ -164,12 +174,15 @@ private void rescale(long now, long next) {
         if (nextScaleTime.compareAndSet(next, now + RESCALE_THRESHOLD)) {
             lockForRescale();
             try {
-                final long oldStartTime = startTime;
-                this.startTime = currentTimeInSeconds();
+                final int oldOffset = startOffset;
+                this.startOffset = currentOffset();
                 final ArrayList<Double> keys = new ArrayList<Double>(values.keySet());
                 for (Double key : keys) {
-                    final Long value = values.remove(key);
-                    values.put(key * exp(-alpha * (startTime - oldStartTime)), value);
+                    final ValueWithOffset value = values.remove(key);
+                    int age = startOffset - value.getOffset();
+                    if (age < maxAge) {
```

Comment:
I realized after writing this that my assumption regarding evictions was untrue -- we only do evictions when we rescale, so it's dependent on the rate at which we remove stale data. If that rate is aggressive, then we get the benefits I described. Otherwise, I think we've got the issue that we're keeping around data that won't be in our snapshot, but which can prevent new values that we might want from being added.